joomla/joomla-cms
# Menu manager for Joomla Backend Menu

Pull Request for issue joomla/joomla-cms#5492

## Summary of Changes
Currently we have menu and menu items for front-end that we manage via Menu Manager. However, for the Backend Menu we do not have any choice. We are bound to use that statically hardcoded menu since long time now.

With this PR I am proposing the flexibility and control over the menu and menu items for Joomla Backend like we have for the Front-end. You can now create, edit, move and organise the menu links the way you like. And if like you can still use the "preset menu" shipped by default with Joomla without any customisation at all.

After this patch you'll have one extra dropdown in Menu manager lists viz. "Client" that can be used to switch between whether you are managing menu and menu items for Front-end or Backend.

When you have client filter "Administrator" selected in the list, clicking on "New" will take you to the Backend Menu Item creation page only. Having "Site" selected will get you a form to create Front-end menu/menu item likewise.  This behavior is certainly not new for us, this can be seen in the existing Module Manager already.

The list of available menu type options are taken the same way that it did for front-end menu type options earlier. All the menu items that exists currently can be created using the Menu Manager in a custom Menu. Then it can be published in the mod_menu for backend, which currently is using a hardcoded menu.

## Notes
* **How will my administrator interface looks change after this?**  
    It will look exactly the same as before. The appearance is only affected if you choose to customise your backend menu via Menu Manager.

* **What do I need to learn to use this new Menu Manager?**  
    Almost Nothing! If you are familiar with the existing Menu Manager and have used it to create menu and menu items for front-end, then you will feel at home. Everything remains the same. When creating a menu item for backend you will see the options related to backend only. Similarly, you will see only front-end options when you create a menu item for front-end. 

* **Where do the menu links gets created for NEW components that I install?**  
    The menu links are created in the same way as before for every new component installation. You can create a "Heading" type menu item (say, "New Components") and set the flag "Components Root" to "Yes". This will pull all those menu items under "New Components".  
    Please note, that this component root is NOT required to be a Top level menu item, it can be any level down.

* **How can I remove an auto created component menu item from the *Component Root* parent after I have created my own menu item for it at desired location?**  
    In the menu items list view, set the menu type filter to "Main (Protected)". You will see all those automatically generated menu items there. Just unpublish the desired items. If the component root has not children to show, the component root will also be hidden automatically.

    Deleting them should not cause any side effect, however its not recommended to do so. Editing these items is disabled to avoid undesirable consequences.   
    
    Please note that if you decide to switch back to the "preset" menu anytime, these auto-generated Menu items will show up under "components" again. Due to B/C reasons preset menu does not obey published/unpublished state.

* **How can I create Multilingual versions of those backend menu items?**
    Language based associations are disabled for backend menu items for consistency and simplicity. The menu title is translated during their rendering, therefore you can use LANGUAGE_KEYS and their translations to use it on multilingual websites. <sup>**1**</sup>

* **Will Joomla! be shipped with a default menu created?**  
    Joomla! will come with built-in preset(s) which will not show up in menu manager. That would be the set as the default menu. You can use it by selecting in the Module manager mod_menu settings. If you want to create and use your own custom menu you can create an new Menu at "Menus > Add New Menu" and build your desired links under it. Next you will choose your own Menu in the Menu module settings via Module Manager.
    
* **What if I do not have a *component root* defined in my custom menu when I install new components?**  
    The "component root" does NOT tell Joomla where to **create** those menu items. It just says, where to **show** them. So relax, the automatic menu items will still be created in the same way as now (even without this PR) irrespective of your *component root* selection. If you have one it will pull them in, and if not they won't show up.  
    This way we can always ensure that they will be shown exactly the same when you switch back to the preset menu for any reason.

* **I messed up my menu structure and it will take time to build it over. Can I get the default Joomla menu back?**  
   Absolutely! Its in there already. Just go to module manager and edit mod_menu for administrator. Choose the *System Preset* as your active menu. Keep it forever, or as long as you need.

* **I am an extension developer. What do I need to change in my components to make use of this new feature?**  
  This update will NOT cause your components to lose any links in the backend menu. If you have any Menu/Submenu under "Components" then your user will still be able to create those links in their custom menu.  
  Well, Ideally you should create layout manifests for all your view/layouts that can be linked to a menu item so as to offer your users better control and flexibility over customisation.

## Testing Instructions
1. Install this patch to the latest 3.7.x branch.
1. You may need to check database and update it from **Extensions** > **Manage** > **Database** and **Fix**. In some cases you may need to manually add `client_id` column to your `#__menu_types` table.  
1. Create a new Menu, say "Admin Menu"
1. Go to Module Manager and Open Menu Module. Choose your new Menu as the Active Menu. The Menu from the top should disappear.
1. Switch back to "System Preset" and the Menu should come back.
1. Make sure the Current Menu (Preset one) is Exactly the same as before this patch. Same access checks should apply.
1. When creating a front-end menu item no backend options should be visible. Same for the other way around.
1. Create a few Menu Items under your "Admin Menu" and then try to set it as your Default Menu in Menu Module. Your created Links should appear.
1. Create a Heading type menu item somewhere in your Menu. Set it as Component Root. All items that came under "Components" earlier should show under this heading now. You may need to publish them from Menu items list first.
1. Install a new component, e.g. "JCE" and see if its menu and sub menu are populated under this component root.
1. Try creating those menu links yourself under some other menu item or even as top level menu item. You should be able to create all those links.
1. Do some more tests around as I may not have listed all possible test cases here.

## Documentation Changes Required
Add information about this feature in the documentation.

Thanks @brianteeman for his support and valuable inputs.


<sup>**1**</sup> <small>In the edit form we can have a set of text inputs for alternative texts per installed language.</small>
